### PR TITLE
transfer bowtie2 from devteam and add interleaved support

### PR DESCRIPTION
**Pull request by nekrut**
- updated to 2.3.2 (new bioconda package created)
- allows for interleaved fastq input
- new test and moved parts of xml to macros

### CODE REVIEW

**Comment by nekrut** [comment]
I think it is ready

File: tools/bowtie2/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+long_description: |
+  Bowtie is an ultrafast and memory-efficient tool for aligning sequencing
+  reads to long reference sequences. It is particularly good at aligning reads of
+  about 50 up to 100s or 1,000s of characters to relatively long (e.g. mammalian)
+  genomes. Bowtie 2 supports gapped, local, and paired-end alignment modes. Bowtie
+  2 outputs alignments in SAM format, enabling interoperation with a large number
+  of other tools.
+name: bowtie2
+owner: devteam
+remote_repository_url: https://github.com/galaxyproject/tools-devteam/tree/master/tools/bowtie2
```
Comment (yhoogstrate) [inline tools/bowtie2/.shed.yml]: tools-devteam or tools-iuc?

File: tools/bowtie2/bowtie2_macros.xml [hunk 2]
```diff
@@ -0,0 +1,323 @@
+<macros>
+    <!-- Import this at the top of your command block and then
+         define rg_auto_name. -->
+    <token name="@define_read_group_helpers@">
+#def identifier_or_name($input1)
+    #if hasattr($input1, 'element_identifier')
+        #return $input1.element_identifier
+    #else
+        #return $input1.name.rstrip('.gz').rstrip('.fastq').rstrip('.fq')
```
Comment (yhoogstrate) [inline tools/bowtie2/bowtie2_macros.xml]: This becomes more complicated with support for GZ and BZ files (e.g. `filename.fq.bz` as given in the test case)

File: tools/bowtie2/bowtie2_macros.xml [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    #else
+        #set $rg_pu = ''
+    #end if
+#end if
+    </token>
+    <token name="@set_use_rg_var@">
+#set $use_rg = str($rg.rg_selector) != "do_not_set"
+    </token>
+    <xml name="read_group_auto_name_conditional">
+        <param name="do_auto_name" type="boolean" label="Auto-assign" help="Use dataset name or collection information to automatically assign this value" checked="no" />
```
Comment (yhoogstrate) [inline tools/bowtie2/bowtie2_macros.xml]: `s/checked="no"/checked="false"/`

File: tools/bowtie2/bowtie2_macros.xml [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                <option value="no" selected="True">No</option>
+                <option value="yes">Yes</option>
+            </param>
+            <when value="yes">
+                <param name="I" type="integer" value="0" min="0" label="Set the minimum fragment length for valid paired-end alignments" help="-I/--minins;  E.g. if `-I 60` is specified and a paired-end alignment consists of two 20-bp alignments in the appropriate orientation with a 20-bp gap between them, that alignment is considered valid (as long as `-X` is also satisfied).  A 19-bp gap would not be valid in that case.  If trimming options `-3` or `-5` are also used, the `-I` constraint is applied with respect to the untrimmed mates. The larger the difference between `-I` and `-X`, the slower Bowtie 2 will run.  This is because larger differences bewteen `-I` and `-X` require that Bowtie 2 scan a larger window to determine if a concordant alignment exists. For typical fragment length ranges (200 to 400 nucleotides), Bowtie 2 is very efficient. Default=0"/>
```
Comment (yhoogstrate) [inline tools/bowtie2/bowtie2_macros.xml]: This can make use of `argument="-I"`

File: tools/bowtie2/bowtie2_macros.xml [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+        <expand macro="read_group_ds_param" />
+        <expand macro="read_group_pi_param" />
+        <expand macro="read_group_dt_param" />
+    </xml>
+    <xml name="read_group_conditional">
+        <conditional name="rg">
+            <param name="rg_selector" type="select" label="Set read groups information?" help="Specifying read group information can greatly simplify your downstream analyses by allowing combining multiple datasets.">
+                <option value="set">Set read groups (SAM/BAM specification)</option>
+                <option value="set_picard">Set read groups (Picard style)</option>
+                <option value="set_id_auto">Automatically assign ID</option>
```
Comment (yhoogstrate) [inline tools/bowtie2/bowtie2_macros.xml]: Automatically assign ID based on name of history item(s)

File: tools/bowtie2/bowtie2_wrapper.xml [hunk 6]
```diff
@@ -0,0 +1,1212 @@
...
+            </param>
+
+            <when value="single">
+                <param name="input_1" format="fastqsanger,fastqsanger.gz,fastqsanger.bz2" type="data" label="FASTQ file" help="Must be of datatype &quot;fastqsanger&quot;" />
+
+                <expand macro="align_unalign" />
+
+            </when>
+            <when value="paired">
+                <param name="input_1" format="fastqsanger,fastqsanger.gz,fastqsanger.bz2" type="data" label="FASTQ file #1" help="Must be of datatype &quot;fastqsanger&quot;" />
```
Comment (yhoogstrate) [inline tools/bowtie2/bowtie2_wrapper.xml]: Would it make sense to create a macro for the input format?

File: tools/bowtie2/bowtie2_wrapper.xml [hunk 7]
```diff
@@ -0,0 +1,1212 @@
...
+                        </action>
+                    </when>
+                </conditional>
+            </actions>
+        </data>
+
+        <data format="bam" name="output" label="${tool.name} on ${on_string}: aligned reads (sorted BAM)">
+          <filter>analysis_type['analysis_type_selector'] == "simple" or analysis_type['sam_opt'] is False</filter>
+          <actions>
+            <conditional name="reference_genome.source">
```
Comment (yhoogstrate) [inline tools/bowtie2/bowtie2_wrapper.xml]: Does someone know if these are set automatically?

File: tools/bowtie2/.shed.yml (hunk 1 above)
Comment (nekrut) [inline tools/bowtie2/.shed.yml]: I believe that to preserve version continuity I need to keep this owner name

Comment (yhoogstrate) [inline tools/bowtie2/.shed.yml]: This is just the URL. In devteam this will move most likely to ./deprecated/

Comment (nsoranzo) [inline tools/bowtie2/.shed.yml]: @nekrut You need to preserve the `owner: devteam` line, but should change the `remote_repository_url: ` to the tools-iuc repo.

**Comment by martenson** [comment]
this is a transfer & update of devteam repohttps://github.com/galaxyproject/tools-devteam/tree/master/tools/bowtie2

we need to create the counterpart PR there and add proper permissions to TTS & MTS

File: tools/bowtie2/.shed.yml (hunk 1 above)
Comment (nekrut) [inline tools/bowtie2/.shed.yml]: done

File: tools/bowtie2/bowtie2_macros.xml (hunk 2 above)
Comment (nekrut) [inline tools/bowtie2/bowtie2_macros.xml]: So should I just update 
```
#return $input1.name.rstrip('.gz').rstrip('.fastq').rstrip('.fq')
```
to 
```
#return $input1.name.rstrip('.gz').rstrip('.fastq').rstrip('.fq').rstrip('.bz2')
```

File: tools/bowtie2/bowtie2_macros.xml (hunk 3 above)
Comment (nekrut) [inline tools/bowtie2/bowtie2_macros.xml]: done

File: tools/bowtie2/bowtie2_macros.xml (hunk 4 above)
Comment (nekrut) [inline tools/bowtie2/bowtie2_macros.xml]: I think the tool was written before `argument` tag existed. I'm a bit lazy to redo all of them

File: tools/bowtie2/bowtie2_wrapper.xml (hunk 6 above)
Comment (nekrut) [inline tools/bowtie2/bowtie2_wrapper.xml]: My issue with this is purely logistical - when things are in the same file it is easier to see things. Obviously, it things are too repetitive it is better to put them away into macros, but this particular case I better keep the way it is.

File: tools/bowtie2/bowtie2_wrapper.xml (hunk 7 above)
Comment (nekrut) [inline tools/bowtie2/bowtie2_wrapper.xml]: I hope so :)

**Comment by martenson** [comment]
permissions guaranteed to the IUC

**Comment by bgruening** [comment]
@yhoogstrate what do you think? Can we merge this?

**Comment by bgruening** [comment]
Not sure, why this PR is testing so much, the bowtie tests seems to be fine.

**Comment by mvdbeek** [comment]
It's green, should we merge?

**Comment by bgruening** [comment]
@mvdbeek think so. 

Imho these kind of PR make it hard to review and see the differences between devteam and IUC. Mabye we can add the original tool into IUC, without modification and blacklist it ... in a follow-up PR we try to enhance it?

**Comment by mvdbeek** [comment]
> Imho these kind of PR make it hard to review and see the differences between devteam and IUC. Mabye we can add the original tool into IUC, without modification and blacklist it ... in a follow-up PR we try to enhance it?

Yeah, I think that would be better.

**Comment by mvdbeek** [comment]
Hmm,
```
cd '/home/travis/build/galaxyproject/tools-iuc/tools/bowtie2' && git rev-parse HEAD
cd '/home/travis/build/galaxyproject/tools-iuc/tools/bowtie2' && git diff --quiet
Could not update bowtie2
Unexpected HTTP status code: 400: {"content_alert": "", "err_msg": "No changes to repository."}
Repository metadata updated successfully for repository 'bowtie2' on the test Tool Shed.
Failed to update repository contents for repository 'bowtie2' on the test Tool Shed.
```
 in https://travis-ci.org/galaxyproject/tools-iuc/jobs/238291862.
That's pretty strange, I'm sure there were changes to the repository.

**Comment by bgruening** [comment]
Its a planemo bug. One of the 4 Travis jobs can upload the others fail I think.

**Comment by nsoranzo** [comment]
It's mainly a bug of our Travis configuration after https://github.com/galaxyproject/tools-iuc/pull/431 was merged, see https://github.com/galaxyproject/tools-iuc/pull/1316 .

**Pull request merged**